### ISSUE TITLE
Can't add new member when has three alive member at four member cluster

### ISSUE THREAD

**Issue by OlivierDeRivoyre** [etcd-io/etcd#10738]
Hello,

I have a cluster with 4 members. I stop the 4th member. I try to add a 5th member to the cluster.
It fails with the error "_not healthy for reconfigure, rejecting member add_" from [server.go#L1582](https://github.com/etcd-io/etcd/blob/master/etcdserver/server.go#L1582).
I expect adding a 5th member works since the three started servers are enough to do a quorum.

Notes: 
 - Done with ETCD version 3.3.11
 - Quite similar to https://github.com/etcd-io/etcd/issues/10499 but without the risk to lose the quorum: [IsReadyToAddNewMember()](https://github.com/etcd-io/etcd/blob/master/etcdserver/api/membership/cluster.go#L554) is ok with 3 members started out of 5.
 - I understand the call to [isConnectedFullySince()](https://github.com/etcd-io/etcd/blob/master/etcdserver/util.go#L47) as a requirement to have all members alive. I suspect the bug to be that isConnectedFullySince() should it be called with only the IsStarted members?

**Comment by jingyih** [comment]
@xiang90 @gyuho Do you remember why `StrictReconfigCheck` requires all members to be connected when adding new member?

**Comment by jingyih** [comment]
@OlivierDeRivoyre My understanding is that 'isConnectedFullySince()` is probably just for extra safety to avoid some corner case. You can bypass the check by setting flag `strict-reconfig-check` to false.

**Comment by OlivierDeRivoyre** [comment]
Hi @jingyih 
The problem with disabling strict-reconfig-check is that it removes the other check IsReadyToAddVotingMember(). This check is relevant and prevents conflicts.

**Comment by hexfusion** [comment]
@OlivierDeRivoyre Can you outline the general use-case for this workflow? I can generally see use but why is a member down and not removed, not sure I understand this? If you can solidify the need for this change it might help us consider the functionality.

**Comment by jingyih** [comment]
`isConnectedFullySince()` is probably added to deal with some corner cases where it is used to prevent a split brain. I suspect it can be relaxed to `the majority is connected` but would require more analyze on possible corner cases.

**Comment by jingyih** [comment]
Did a little digging, `isConnectedFullySince()` is introduced by #6106 to fix #6103. In short, I think it sacrifices the availability of member add API to prevent potential user error. When a member in 3-member-cluster is down (but "started" [1]) and needed to be replaced, the recommended way is to first remove it, then add a new one. If user ignores the recommendation and try to first add a member then remove the failed old member, cluster might lose quorum if the new member is mis-configured and won't start. `isConnectedFullySince()` is necessary to prevent this kind of user error for the described 3-member-cluster case, but too strict for a 5-member-cluster.

[1] If a member is started then failed (due to hardware failure, network failure, etc), the member is still considered "started".

**Comment by dlipovetsky** [comment]
I want to share a different example which I think may share the root cause.

> For avoiding this problem, etcd provides an option -strict-reconfig-check. If this option is passed to etcd, etcd rejects reconfiguration requests if the number of started members will be less than a quorum of the reconfigured cluster.

According to the [documentation](https://etcd.io/docs/v3.3.12/op-guide/runtime-configuration/), it should be possible to concurrently add two members to a five member cluster. With five members, the quorum is three. When the sixth member is added, the quorum becomes four. When the seventh is added, the quorum remains four.

Here is a cluster with five started, healthy members. I can add the sixth member, but cannot add a seventh member (until the sixth member is started):
```
$ etcdctl member list
187096471e8b9fe2, started, isv-daniel-844-10-105-16-53platform9.sys, https://10.105.16.53:2380, https://10.105.16.53:2379
2467199df27fb673, started, isv-TeamCity-847-10-105-16-76platform9.sys, https://10.105.16.76:2380, https://10.105.16.76:2379
57c5b6d4bd41268b, started, isv-daniel-844-10-105-16-98platform9.sys, https://10.105.16.98:2380, https://10.105.16.98:2379
5ef1c5c69d8de05e, started, isv-TeamCity-847-10-105-16-46platform9.sys, https://10.105.16.46:2380, https://10.105.16.46:2379
90eb2b94452b46f2, started, isv-daniel-844-10-105-16-85platform9.sys, https://10.105.16.85:2380, https://10.105.16.85:2379
$ etcdctl member add member6 --peer-urls https://1.1.1.6:2380
Member 112188a454b533bf added to cluster e4a359f6dfce2e7b

ETCD_NAME="member6"
ETCD_INITIAL_CLUSTER="member6=https://1.1.1.6:2380,isv-daniel-844-10-105-16-53platform9.sys=https://10.105.16.53:2380,isv-TeamCity-847-10-105-16-76platform9.sys=https://10.105.16.76:2380,isv-daniel-844-10-105-16-98platform9.sys=https://10.105.16.98:2380,isv-TeamCity-847-10-105-16-46platform9.sys=https://10.105.16.46:2380,isv-daniel-844-10-105-16-85platform9.sys=https://10.105.16.85:2380"
ETCD_INITIAL_ADVERTISE_PEER_URLS="https://1.1.1.6:2380"
ETCD_INITIAL_CLUSTER_STATE="existing"
$ etcdctl member list
112188a454b533bf, unstarted, , https://1.1.1.6:2380, 
187096471e8b9fe2, started, isv-daniel-844-10-105-16-53platform9.sys, https://10.105.16.53:2380, https://10.105.16.53:2379
2467199df27fb673, started, isv-TeamCity-847-10-105-16-76platform9.sys, https://10.105.16.76:2380, https://10.105.16.76:2379
57c5b6d4bd41268b, started, isv-daniel-844-10-105-16-98platform9.sys, https://10.105.16.98:2380, https://10.105.16.98:2379
5ef1c5c69d8de05e, started, isv-TeamCity-847-10-105-16-46platform9.sys, https://10.105.16.46:2380, https://10.105.16.46:2379
90eb2b94452b46f2, started, isv-daniel-844-10-105-16-85platform9.sys, https://10.105.16.85:2380, https://10.105.16.85:2379
$ etcdctl member add member7 --peer-urls https://1.1.1.7:2380
Error: etcdserver: unhealthy cluster
$ etcdctl endpoint health
127.0.0.1:2379 is healthy: successfully committed proposal: took = 3.035203ms

```

It looks like etcd is serializing member adds, something more strict than what the documentation describes.

However, serializing member adds is actually _desirable_ when writing tooling that dynamically bootstraps an etcd cluster. If etcd does _not_ serialize member adds, then once it you have five started members, you can add two members concurrently. If two members are _added_ concurrently, then one member may have a stale member list when it tries to _join_.

To recap, I think that the original example reveals a bug.
> I have a cluster with 4 members. I stop the 4th member. I try to add a 5th member to the cluster.
> I expect adding a 5th member works since the three started servers are enough to do a quorum.
It should be possible to add the 5th member in this case. There are no unstarted members, and there is a quorum.

However, I think my example reveals behavior (namely, serialized member adds) that I think is desirable, might be seen as a bug because it contradicts the documentation. I hope that, as we fix the above bug, we preserve the serialized member adds and update the documentation.

**Comment by jingyih** [comment]
@dlipovetsky Appreciate your observation and analysis. This is a valid point, I agree the membership reconfiguration API should preserve the behavior you mentioned.
